File: src/jax_translation/symmetric_polynomials.jl
module SymmetricPolynomials
export get_symmetric_polynomials!
"""
    get_symmetric_polynomials!(dest, roots, b::Int, reg_coeffs)
Calculate the first `b` elementary symmetric polynomials for the given roots regularized by ∏ᵢreg_coeffs[i].

# Arguments
- `dest`: Destination array for the computed elementary symmetric polynomials
- `roots`: Vector of elements for which the elementary symmetric polynomials are to be computed
- `b::Int`: The maximum order of elementary symmetric polynomials to compute.
- `reg_coeffs`: Array of regularization coefficients.

# Returns
- `nothing`: In-place computation
"""
function get_symmetric_polynomials!(dest, roots, b, reg_coeffs)

    dest[1] = one(eltype(dest))

    if b == 0
        return
    elseif b == 1
        dest[2] = sum(roots) * reg_coeffs[1] ## okay.
        return
    end

    dest[2:end] .= zero(eltype(dest))

    for i in eachindex(roots)
        r = roots[i]
        for j in min(i, b):-1:1

            @inbounds dest[j + 1] += r * dest[j] * reg_coeffs[j]

        end

    end

    return

end
end

File: src/jax_translation/symmetric_polynomials.py
import jax.numpy as jnp
from jax import jit, vmap
from functools import partial

@partial(jit, static_argnums=(1,))
def get_symmetric_polynomials_single(roots, b, reg_coeffs):
    """
    Calculate the first `b` elementary symmetric polynomials for a single set of roots,
    regularized by ∏ᵢreg_coeffs[i].

    Args:
        roots: Array of elements for which the elementary symmetric polynomials are to be computed.
        b: The maximum order of elementary symmetric polynomials to compute.
        reg_coeffs: Array of regularization coefficients.

    Returns:
        A JAX array containing the computed elementary symmetric polynomials.
    """
    dest = jnp.zeros(b + 1, dtype=roots.dtype)
    dest = dest.at[0].set(1)  # Set the first element to 1

    if b == 0:
        return dest
    elif b == 1:
        dest = dest.at[1].set(jnp.sum(roots) * reg_coeffs[0])
        return dest

    for i, r in enumerate(roots):
        for j in range(min(i + 1, b), 0, -1):
            dest = dest.at[j].add(r * dest[j - 1] * reg_coeffs[j - 1])

    return dest

# Vectorized version of the function
@partial(jit, static_argnums=(1,))
def get_symmetric_polynomials(roots_matrix, b, reg_coeffs):
    """
    Vectorized version of get_symmetric_polynomials_single to handle a batch of roots.

    Args:
        roots_matrix: A matrix of shape (B, N), where B is the batch size and N is the number of roots.
        b: The maximum order of elementary symmetric polynomials to compute.
        reg_coeffs: Array of regularization coefficients.

    Returns:
        A JAX array of shape (B, b + 1) containing the computed symmetric polynomials for each batch.
    """
    return vmap(lambda roots: get_symmetric_polynomials_single(roots, b, reg_coeffs))(roots_matrix)